### PR TITLE
feat(mpt): Block hash walkback

### PR DESCRIPTION
**Pull request by clabby**
## Overview

Introduces the implementation for `Database::block_hash` for `TrieDB`, allowing the `BLOCKHASH` opcode to verifiably fetch a block hash within the most recent `256` blocks, not including the block that is currently being executed.

### CODE REVIEW

**Comment by clabby** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/ethereum-optimism/kona/199?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#199** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/kona/199?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#198** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/kona/198?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#197** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/kona/197?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#196** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/kona/196?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#195** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/kona/195?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @clabby and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: crates/mpt/src/db/mod.rs [hunk 1]
```diff
@@ -307,11 +352,27 @@ where
...
-        //     Entry::Vacant(_) => anyhow::bail!("Block hash for number not found"),
-        // }
-        unimplemented!("Block hash not implemented; Need to unroll the starting block hash for this operation.")
+    fn block_hash(&mut self, block_number: U256) -> Result<B256, Self::Error> {
+        // The block number is guaranteed to be within the range of a u64.
+        let u64_block_number: u64 = block_number.to();
+
+        // Fetch the block header from the preimage fetcher.
+        let mut block_hash = self.parent_block_hash;
+        let mut header = (self.header_by_hash_fetcher)(block_hash)?;
```
Comment (refcell) [inline crates/mpt/src/db/mod.rs:361]: Do we need these parenthesis?

```suggestion
        let mut header = self.header_by_hash_fetcher(block_hash)?;
```

File: crates/mpt/src/db/mod.rs [hunk 2]
```diff
@@ -307,11 +352,27 @@ where
         Ok(int_slot)
     }
 
-    fn block_hash(&mut self, _: U256) -> Result<B256, Self::Error> {
-        // match self.db.block_hashes.entry(number) {
-        //     Entry::Occupied(entry) => Ok(*entry.get()),
-        //     Entry::Vacant(_) => anyhow::bail!("Block hash for number not found"),
-        // }
-        unimplemented!("Block hash not implemented; Need to unroll the starting block hash for this operation.")
+    fn block_hash(&mut self, block_number: U256) -> Result<B256, Self::Error> {
```
Comment (refcell) [inline crates/mpt/src/db/mod.rs:355]: There's no way to do this without walking back huh?

File: crates/mpt/src/db/mod.rs (hunk 1 above)
Comment (clabby) [reply crates/mpt/src/db/mod.rs:361]: Yep, we do. Rust won't compile w/o it.

File: crates/mpt/src/db/mod.rs (hunk 2 above)
Comment (clabby) [reply crates/mpt/src/db/mod.rs:355]: Yep, that's the way to do it in a verifiable way, starting from the latest parent block hash. In a normal client, you could just consult the DB, but with the execution for this DB being stateless, we potentially need to verifiably fetch data prior to the initialization point (i.e., in the first block we execute, there's a `BLOCKHASH <current_block> - 200` opcode.

**Review by refcell — APPROVED** [review]

**Pull request merged**